File: SAMPL/sourceCode/SAMPLcore/Physics/ComputeTransferMatrix.py
import numpy as np


def ComputeTransferMatrix(beamline, startStop, beam, trajectory):
    n = startStop[1] - startStop[0] + 2
    eref = beam.energy * np.ones(n)
    m = np.zeros((n, 6, 6))
    m[0, :, :] = np.eye(6)

    precn = beamline.precision
    print precn
    p = precn * np.eye(6, 6)
    for i in range(6):
        p[:, i] = p[:, i] + trajectory

    beam.particles = p
    #print beam.particles

    for i in range(startStop[0], startStop[1]):
        beam = beamline.TrackMatlab([i, i], beam)
        p1 = beam.particles
        print p1
        for j in range(6):
            m[i - startStop[0] + 2, j, :] = (p1[j] - trajectory) / precn

        eref[i - startStop[0] + 2] = beam.energy

    return m, eref

File: SAMM3.0/Physics/ComputeTransferMatrix.m
function [m, eref] = ComputeTransferMatrix(beamline,range,beam,trajectory)

    eref     = beam.energy * ones(1,range(2)-range(1)+2);
    m        = zeros(6,6,range(2)-range(1)+2);
    m(:,:,1) = eye(6);

    precn = beamline.precision;
    p  = precn*eye(6,7);
    for n = 1:7
        p(:,n) = p(:,n) + trajectory;
    end

    beam.particles = p;
    p2 = p(:,7)

    for n = range(1):range(2)
        beam = beamline.Track([n n],beam);
        p1 = beam.particles
        for v = 1:6
            m(:,v,n-range(1)+2) = (p1(:,v) - p1(:,7)) / precn;
        end
        eref(n-range(1)+2) = beam.energy;
    end

return
